### DISCUSSION TITLE
递归DTO存在一些问题

### DISCUSSION THREAD

**Discussion by lyric-zemin** [midwayjs/midway#4254] [Q&A]
### Describe the problem（描述问题）

```ts
import { getSchema, Rule, RuleType } from '@midwayjs/validate';

// 第一种写法导致children字段被剔除
export class MenuItemDTO {
  @Rule(RuleType.number())
  id?: number;

  @Rule(RuleType.string().required())
  name: string;

  @Rule(RuleType.string().required())
  path: string;

  // BUG 这里没拿到rule Schema
  @Rule(RuleType.array().items(getSchema(MenuItemDTO)))
  children: MenuItemDTO[];
}

// 第二中写法导致第二级的children字段被剔除
export class MenuDTO extends MenuItemDTO {
  // BUG 这里的rule 没有children字段
  @Rule(RuleType.array().items(getSchema(MenuItemDTO)))
  children: MenuItemDTO[];
}

```

### Midway Versions（Midway 版本）

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.19.0)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/busboy(3.19.2)
✓ @midwayjs/cache-manager(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.19.0)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(not installed)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(3.19.2)
✓ @midwayjs/info(not installed)
✓ @midwayjs/jwt(3.19.2)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.19.2)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(3.19.2)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(not installed)
✓ @midwayjs/security(3.19.2)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(3.19.2)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(3.19.2)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.19.2)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.19.2)
✓ @midwayjs/ws(not installed)

**Comment by czy88840616** [comment]
递归。。。。

**Reply by lyric-zemin** [reply]
貌似这种情况只能使用自定义校验了

**Reply by DannyWin** [reply]
@czy88840616 @lyric-zemin 
遇到同样问题了，这种需求比较普遍，比如管理系统中的菜单就类似这种结构，以下是解决方案（children可以触发级联校验）

import { getSchema, Rule, RuleType } from '@midwayjs/validate';

export class MenuItemDTO {
  @Rule(RuleType.number())
  id?: number;

  @Rule(RuleType.string().required())
  name: string;

  @Rule(RuleType.string().required())
  path: string;

  @Rule(RuleType.array().items(**RuleType.link('/')**))
  //或者使用@Rule(RuleType.array().items(**RuleType.link('...')**))
  children: MenuItemDTO[];
}
依据:https://joi.dev/api/?v=17.13.3#linkref
